rpc: Make unloadwallet wait for complete wallet unload

promag opened the pull request:
Currently the `unloadwallet` RPC is asynchronous, it only signals the intent to unload the wallet and then returns the response to the client. The actual unload can happen later and the client has no way to be notified of that.

This PR makes the `unloadwallet` RPC synchronous, meaning that it blocks until the wallet is fully unloaded.

Replaces #14919, fixes #14917.

promag committed:
rpc: Make unloadwallet wait for complete wallet unload

promag commented:
@ryanofsky @MarcoFalke

laanwj commented:
Concept ACK - I think this can be useful outside of the tests, as well.

> Also I'd like to discuss the wait default value, currently it avoid changing behavior. If wait=true is considered the correct behavior then this could be backport to 0.17.2?

Strictly spoken this is a feature, not a bugfix, though if it's optional and improves testing I think a point can be made to backport it anyhow.

ryanofsky commented:
Code change looks ok, but this shouldn't be an option. Making `wait=false` the default makes `unloadwallet` pointlessly difficult to use, and inconsistent with other methods, including closely related methods like `createwallet` and `loadwallet`.

If you think there is reason to support `wait=false`, I think you need to say what the use-case is, and why it wouldn't be better served by having the client make a fully asynchronous JSON-RPC request, instead of using this half-asynchronous/half-synchronous `wait` option that doesn't (I guess?) block on acquiring `cs_wallet`, but does block on the network connection and on acquiring `cs_wallets`, and on whatever the UI currently does in `NotifyUnload` (or will do in the future).

Also, if you want to keep the `async=false` option, it would be good to have a test to ensure that it doesn't acquire `cs_wallet` and `cs_main`. This would be easy to write as a c++ unit test which acquires the locks in one thread and calls `unloadwallet` in another.

promag commented:
Changed to synchronous unload. Added some sleeps to see if there were connection timeout, but didn't manage to cause one.

> doesn't (I guess?) block on acquiring `cs_wallet`

no it doesn't, it waits until the weak pointer is expired. An alternative is to keep the shared pointer to force the unload in the RPC thread.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -81,13 +81,33 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
 }
 
+static std::mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cond;
```
It seems more common in bitcoin code to abbreviate condition variables as `cv` instead of `cond` and name locks `lock` instead of `lk`.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -81,13 +81,33 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+        delete wallet;
+    }
+
+    g_wallet_release_cond.notify_all();
```
I think we need to confirm that `!weak_wallet.expired()` in the other thread is true at the point of this call (and guaranteed by the c++ standard to be true). Otherwise there could be a lost-wakeup bug here because this would be signalling the condition variable before the condition is true.

promag replied:
Good catch, I'll investigate.

promag replied:
This no longer applies, @ryanofsky concern was valid.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -81,13 +81,33 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+    wallet.reset();
+
+    std::unique_lock<std::mutex> lk(g_wallet_release_mutex);
+    while (!weak_wallet.expired()) {
```
It looks like this might not be sufficient if we want to guarantee that the wallet is really destroyed. From https://en.cppreference.com/w/cpp/memory/weak_ptr/expired:

> The destructor for the managed object may not yet have been called, but this object's destruction is imminent (or may have already happened).

A more reliable but cumbersome approach might be to add wallet members:

```c++
std::promise<void> m_unload_promise;
std:shared_future<void> m_unload_future = m_unload_promise.get_future();
```

and call `m_unloaded_promise.set_value()` in the destructor, and then `future.wait()` here.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -81,13 +81,46 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+    }
+    // Now it's safe to close and delete the wallet.
+    pwallet->WalletLogPrintf("Releasing wallet\n");
+    pwallet->BlockUntilSyncedToCurrentChain();
```
Not sure about this, I think it should be called before `UnregisterValidationInterface` otherwise `SyncWithValidationInterfaceQueue` does nothing than waiting..

ryanofsky replied:
This does seem unnecessary. But I'd also expect this code to calling UnregisterValidationInterface before flushing. I.e. stop requesting new notifications, then save state, instead of saving state first and then hoping new notifications don't arrive after the flush...

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -82,13 +82,46 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+    {
+        std::unique_lock<std::mutex> lock(g_wallet_release_mutex);
+        // Assert that there was an explict UnloadWallet call.
+        assert(g_unloading_wallet_set.erase(wallet) == 1);
```
Would suggest:

```c++
size_t erased = g_unloading_wallet_set.erase(wallet) == 1
assert(erased == 1);
```

Even though we don't support building bitcoin with NDEBUG, I think most people just expect asserts to contain error checking, not code that is important to run.

ryanofsky reviewed (commented):
This change seems like it should work, but multiwallet test is failing. There's some problem with combine_logs on travis, but locally I see:

```
AssertionError: [node 0] Expected message "BerkeleyBatch: Can't open database w8_copy \(duplicates fileid \w+ from w8\)" does not partially match stderr:
"bitcoind: wallet/wallet.cpp:95: void ReleaseWallet(CWallet*): Assertion `g_unloading_wallet_set.erase(wallet) == 1' failed."
```

promag commented:
@ryanofsky looks you started reviewing before my latest push?

promag commented on src/wallet/wallet.cpp:
```diff
@@ -82,15 +82,59 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+        unloading = g_unloading_wallet_set.erase(wallet) == 1;
+    }
+    if (unloading) {
+        // Delegate wallet deletion for UnloadWallet caller.
```
`Delegate wallet deletion to UnloadWallet caller`

promag commented on src/wallet/wallet.cpp:
```diff
@@ -82,15 +82,59 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+        // Delegate wallet deletion for UnloadWallet caller.
+        g_wallet_release_cv.notify_all();
+    } else {
+        // Delegate wallet deletion for UnloadWallet caller.
```
Could say it's not necessary to call `NotifyUnload` because at this point there are no shared pointers to this wallet.

promag reviewed (commented):
Self review, 2 comments must be fixed.

promag commented:
@ryanofsky would appreciate another review! 😄ty

gmaxwell commented:
Concept ACK.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -82,15 +82,63 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+// Custom deleter for shared_ptr<CWallet>.
+static void ReleaseWallet(CWallet* wallet)
+{
+    bool unloading{false};
```
Is there a reason this needs to delegate deletion? It seems if you removed the DeleteWallet call from UnloadWallet, you could simplify this function to just:

```c++
DeleteWallet(wallet);
LOCK(g_wallet_release_mutex);
g_unloading_wallet_set.erase(wallet);
g_wallet_release_cv.notify_all();
```

promag replied:
No strong reason, just thought the unload burden/time should be left to the `unloadwallet` caller. The way you suggest both the `unloadwallet` caller and the last wallet client wait for the complete unload.

ryanofsky replied:
> just thought the unload burden/time should be left to the unloadwallet caller. 

Interesting. That seems like a legitimate optimization for the last RPC using the wallet. Personally, I think I'd still choose the simplification over the optimization, but if you want to keep the optimization it'd be good to have a comment saying why it's better to delete on the unload thread than the release thread.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -54,6 +54,10 @@ void StopWallets();
 
+//! Unloads and deletes the wallet.
+//  Blocks current thread until the wallet is deleted.
+void UnloadWallet(std::shared_ptr<CWallet>&& wallet);
```
Comment could say that it's not necessary to call this function unless you need to block waiting the unload.

promag replied:
Will improve comment.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -82,15 +82,63 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+    // The wallet can be in use so it's not possible to explicitly unload here.
+    // Notify the unload intent so that all remaining shared pointers are
+    // released.
+    pwallet->NotifyUnload();
```
Could we `assert(NotifyUnload.empty())` in the wallet destructor to verify delete isn't happening too early and clients are fully detached?

promag replied:
Will try.

ryanofsky reviewed (approved):
utACK 7aed2bb1f06ec505507d28f623b4973ae314c3f. This seems right but I suggested a simplification. It would also be good to update the description and add a release note.

promag committed:
doc: Add release notes for unloadwallet change to synchronous call

promag commented:
Updated after @ryanofsky review, added release notes and updated description.

gmaxwell commented:
Concept ACK.

ryanofsky reviewed (approved):
utACK c2ae9d9c295ca75f6653892b5f5e1ad6f5c1a9e7. Changes since last review were adding documentation and assert. Also fixing unrelated (I think) encrypted_batch potential memory leak in CWallet destructor.

laanwj commented:
utACK c2ae9d9c295ca75f6653892b5f5e1ad6f5c1a9e7

jnewbery commented:
Lightly tested ACK, although I have a preference for @ryanofsky's proposed simplification here: https://github.com/bitcoin/bitcoin/pull/14941#discussion_r245133485

promag commented:
Pushed the simplification, also split code from assertions.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -82,15 +82,56 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
 }
 
+// Custom deleter for shared_ptr<CWallet>.
+// If the release is due to a UnloadWallet call then deletion is delegated to
```
This comment needs to be updated with the latest change.

Also nit: s/a UnloadWallet/an UnloadWallet/

jnewbery reviewed (commented):
Looks good. Comment needs updating, and I'd squash the new commit 'fixup: Russ simplification' into 'rpc: Make unloadwallet wait for complete wallet unload'

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -102,19 +102,11 @@ static void DeleteWallet(CWallet* wallet)
-        // more shared pointers to the wallet.
-        DeleteWallet(wallet);
-    }
+    DeleteWallet(wallet);
```
Would suggest inlining `DeleteWallet` here. This function is only called from this one place, and I can't think of any other place where it would be safe to call.

promag replied:
Good point.

ryanofsky reviewed (approved):
utACK e6cfc2b566b5a036d884754363b95ac94808907d. Only change since last review is new commit.

promag commented:
Latest changes in 4a50dce.

jnewbery commented:
utACK 645e905c327411555073fa7964b36f652998059f
